torch/csrc/jit/script/class_type.cpp

```diff
@@ -110,6 +110,17 @@ ClassType::ClassType(
 bool ClassType::isSubtypeOfExt(const TypePtr rhs, std::ostream* why_not) const {
   // to improve performance, this check can be cached
   if (auto iface = rhs->cast<InterfaceType>()) {
+    // ClassType is not a subtype of InterfaceType if the InterfaceType is a
+    // Module Interface Type but the Class Type is not a Module Class Type
+    if(!is_module() && iface->is_module()) {
```

Comment:
In Python, doing `mymod.foo = SomeModule` automatically registers a submodule, so you can't normally get into a situation where an `nn.Module` behaves as a normal class attribute. So I think we should restrict this to avoid confusion.

I don't think the check should be in `setattr`—`setattr` shouldn't do too much special stuff. We should locate the check in this part of the code, since it's responsible for deciding "does object X fulfill interface Y".

Reply:
I actually disagree here. A module with is a more specific type of a class. If it has the appropriate methods to match a non-Module interface, then it fulfills the contract of that interface and should be a subtype of that interface. Let's say someone has an interface with a `str` method on it. If the module defines `str` then it should be able to be passed as a member of that interface even if it is not itself an interface. The reverse is not true, only module instances can be members of a module interface.